### ISSUE TITLE
potential deadlock with scx_flatcg

### ISSUE THREAD

**Issue by arighi** [sched-ext/scx#49]
scx_flatcg is calling  `bpf_cgroup_from_id()` from `.dispatch()`, that is called with `rq->__lock` held, `bpf_cgroup_from_id()` is basically a kfunc wrapper of `cgroup_get_from_id()` that is calling `kernfs_find_and_get_node_by_id()`, that can acquire `kernfs_idr_lock` (with `rq->_lock` held).

In kernfs we can acquire `kernfs_idr_lock`, be interrupted by a scheduler tick and acquire `rq->__lock`, leading to the following deadlock scenario:
```
        CPU0                    CPU1
        ----                    ----
   lock(kernfs_idr_lock);
                                lock(rq->__lock);
                                lock(kernfs_idr_lock);
   <Interrupt>
    lock(rq->__lock);
```

I don't know exactly how to prevent this without doing a big redesign of scx_flatcg or changing kernfs to use irqsave/irqrestore spinlocks (that seems a pretty strong change... but probably the safest if we want to make `bpf_cgroup_from_id()` available for general usage).

Suggestions?

**Comment by htejun** [comment]
Yeah, the right thing to do is making kernfs_idr_lock irqsafe and that shouldn't make anything noticeably worse. Hmm... given that bpf_cgroup_from_id() is a kfunc which doesn't have any restrictions, it could be that this deadlock is already possible with other BPF progs. If so, it's an easy patch to justify. I'll ask BPF folks whether that's already the case.

**Comment by arighi** [comment]
FWIW, I did a test with this applied and everything looks good: https://github.com/sched-ext/sched_ext/compare/sched_ext...arighi:sched_ext:kernfs-irq-spinlock

I was able to trigger the lockdep splat running scx_flatcg inside virtme-ng:
```
vng -v -- ./build/scheds/c/scx_flatcg
```

BTW I'm also working on a github workflow to test the schedulers automatically using virtme-ng, that should help to catch issues like this in advance, I may be able to send a PR later today or tomorrow.

**Comment by htejun** [comment]
We'll probably have to use raw_spinlock given that it can be nested inside rq lock. BPF folks say it should be possible to trigger the same condition without scx given that the kfunc can be called from any kprobe/tracepoint/fentry attached BPF progs.

**Comment by htejun** [comment]
This issue should be fixed by: https://github.com/sched-ext/sched_ext/commit/1225a9069ff2185007b078b7e447727d229a64fe
